File: scripts/check-frontend-config-cache.py
#!/usr/bin/env python3
import argparse
import os
import re
import sys


def strip_strings(line: str) -> str:
    line = re.sub(r'"(?:[^"\\]|\\.)*"', '""', line)
    line = re.sub(r"'(?:[^'\\]|\\.)*'", "''", line)
    line = re.sub(r'`(?:[^`\\]|\\.)*`', '``', line)
    return line


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('--repo-root', default=os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
    parser.add_argument('--fail-on-finding', action='store_true')
    args = parser.parse_args()

    repo_root = os.path.abspath(args.repo_root)
    frontend_src = os.path.join(repo_root, 'frontend', 'src')
    if not os.path.isdir(frontend_src):
        print('WARN: frontend/src not found; skipping frontend config cache scan.')
        return 0

    pattern = re.compile(r'^\s*(const|let)\s+[A-Za-z0-9_]+\s*=\s*(getBackendUrl|getAssistUrl)\(\)')
    findings: list[str] = []

    for root, _, files in os.walk(frontend_src):
        for name in files:
            if not name.endswith(('.ts', '.tsx')):
                continue
            file_path = os.path.join(root, name)
            brace_depth = 0
            with open(file_path, encoding='utf-8') as handle:
                for index, line in enumerate(handle, start=1):
                    if brace_depth == 0 and pattern.search(line):
                        relative = os.path.relpath(file_path, repo_root).replace(os.sep, '/')
                        findings.append(f'{relative}:{index}: {line.strip()}')

                    cleaned = strip_strings(line)
                    brace_depth = max(0, brace_depth + cleaned.count('{') - cleaned.count('}'))

    if not findings:
        print('OK: no obvious module-scope frontend URL caches found.')
        return 0

    print("WARN: possible frontend URL caches found. Prefer call-time helpers like apiBase() => getBackendUrl() || ''.")
    for finding in findings:
        print(f'  {finding}')

    return 1 if args.fail_on_finding else 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/check-frontend-config-cache.mjs
#!/usr/bin/env node
import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { fileURLToPath } from 'node:url'

const scriptDir = path.dirname(fileURLToPath(import.meta.url))
let repoRoot = path.resolve(scriptDir, '..')
let failOnFinding = false

const args = process.argv.slice(2)
for (let index = 0; index < args.length; index += 1) {
  const arg = args[index]
  if (arg === '--repo-root') {
    repoRoot = path.resolve(args[index + 1])
    index += 1
  } else if (arg === '--fail-on-finding') {
    failOnFinding = true
  } else {
    console.error(`Unknown argument: ${arg}`)
    process.exit(2)
  }
}

const frontendSrc = path.join(repoRoot, 'frontend/src')
if (!fs.existsSync(frontendSrc)) {
  console.log('WARN: frontend/src not found; skipping frontend config cache scan.')
  process.exit(0)
}

const findings = []
const declarationPattern = /^\s*(const|let)\s+[A-Za-z0-9_]+\s*=\s*(getBackendUrl|getAssistUrl)\(\)/

function walk(dir) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const entryPath = path.join(dir, entry.name)
    if (entry.isDirectory()) {
      walk(entryPath)
    } else if (/\.(ts|tsx)$/.test(entry.name)) {
      scanFile(entryPath)
    }
  }
}

function stripStrings(line) {
  return line
    .replace(/"(?:[^"\\]|\\.)*"/g, '""')
    .replace(/'(?:[^'\\]|\\.)*'/g, "''")
    .replace(/`(?:[^`\\]|\\.)*`/g, '``')
}

function scanFile(filePath) {
  const lines = fs.readFileSync(filePath, 'utf8').split(/\r?\n/)
  let braceDepth = 0

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index]
    if (braceDepth === 0 && declarationPattern.test(line)) {
      const relative = path.relative(repoRoot, filePath).replace(/\\/g, '/')
      findings.push(`${relative}:${index + 1}: ${line.trim()}`)
    }

    const withoutStrings = stripStrings(line)
    const opens = [...withoutStrings.matchAll(/\{/g)].length
    const closes = [...withoutStrings.matchAll(/\}/g)].length
    braceDepth = Math.max(0, braceDepth + opens - closes)
  }
}

walk(frontendSrc)

if (findings.length === 0) {
  console.log('OK: no obvious module-scope frontend URL caches found.')
  process.exit(0)
}

console.log("WARN: possible frontend URL caches found. Prefer call-time helpers like apiBase() => getBackendUrl() || ''.")
for (const finding of findings) {
  console.log(`  ${finding}`)
}

process.exit(failOnFinding ? 1 : 0)